packages/foundry/contracts/test/other/tether/TetherJoin.t.sol

```diff
@@ -0,0 +1,285 @@
+        vm.prank(ladle);
+        join.join(user, unit);
+
+        assertTrackMinusEq("userBalance", unit, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", unit, join.storedBalance());
+        assertTrackPlusEq("joinBalance", unit, tether.balanceOf(address(join)));
+    }
+}
+
+abstract contract WithFees is Deployed {
+    function setUp() public override virtual {
+        super.setUp();
+
+        // enable fees
+        vm.prank(tetherMultiSig);
+        tether.setParams(19, 49);    // maximum
+    }
+}
+
+contract WithFeesTest is WithFees {
+    function testJoinPullWithFees() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        uint256 units = unit * 5;
+        uint256 fee = tether.basisPointsRate() * 100;
+        uint256 feeAdjustedUnits = units * unit / (unit - fee); // scale up first so precision isn't lost
+
+        vm.prank(user);
+        tether.approve(address(join), feeAdjustedUnits);
+        vm.prank(ladle);
+        join.join(user, uint128(feeAdjustedUnits));
+
+        assertTrackMinusEq("userBalance", feeAdjustedUnits, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", units, join.storedBalance());
+        assertTrackPlusEq("joinBalance", units, tether.balanceOf(address(join)));
+    }
+
+    function testJoinPushWithFees() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        uint256 units = unit * 5;
+        uint256 fee = tether.basisPointsRate() * 100;
+        uint256 feeAdjustedUnits = units * unit / (unit - fee); // scale up first so precision isn't lost
+
+        vm.prank(user);
+        tether.transfer(address(join), feeAdjustedUnits);
+        vm.prank(ladle);
+        join.join(user, uint128(units));
+
+        assertTrackMinusEq("userBalance", feeAdjustedUnits, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", units, join.storedBalance());
+        assertTrackPlusEq("joinBalance", units, tether.balanceOf(address(join)));
+    }
+
+    function testJoinCombineWithFees() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        uint256 units = unit * 13;
+        uint256 fee = tether.basisPointsRate() * 100;
+        uint256 feeAdjustedUnits = units / 2 * unit / (unit - fee); // scale up first so precision isn't lost
+
+        vm.prank(user);
+        tether.approve(address(join), feeAdjustedUnits);
+        vm.prank(user);
+        tether.transfer(address(join), feeAdjustedUnits);
+        vm.prank(ladle);
+        join.join(user, uint128(units));
+
+        assertTrackMinusEq("userBalance", feeAdjustedUnits * 2, tether.balanceOf(user));
```

Comment:
Here your asserts are more complicated. Let's make a test where the user spends `amount * 2`.

 - Have the user to send an `amount` and obtain `amountReceived` by checking the change on `tether.balanceOf(address(join))`. You also obtain `feeCharged` as the difference.
 - Ask for  `join.join(amountReceived * 2)`. Since the `join` pulls to receive exactly what it is told, and it already has `amountReceived`, it will aim to obtain another `amountReceived` and the reverse fee formula should mean the user transfers `amount`.

Now you can check:
 - The user tether balance dropped by `amount * 2`.
 - The join tether balance increased by `amountReceived * 2`
 - The join stored balance increased by `amountReceived * 2`

This is also a very nice test, because we get to test our reverse fee implementation against tether's fee implementation, with no assumptions on the tests as to what the formulas actually are.